crates/ty_python_semantic/src/types/constraints.rs

```diff
@@ -273,1231 +239,1147 @@ impl<'db> ConstraintSet<'db> {
-
-        // If the requested constraint is `¬(Never ≤ T ≤ object)`, then the constraint cannot be
-        // satisfied.
-        let negative = NegatedRangeConstraint {
-            hole: RangeConstraint { lower, upper },
-        };
-        if negative.hole.is_always() {
-            return Satisfiable::Never;
-        }
-
-        Satisfiable::Constrained(Constraint {
-            positive: RangeConstraint::always(),
-            negative: smallvec![negative],
-        })
-    }
-}
-
-impl<'db> NegatedRangeConstraint<'db> {
-    /// Clips this negative hole to be the smallest hole that removes the same types from the given
-    /// positive range.
-    fn clip_to_positive(&self, db: &'db dyn Db, positive: &RangeConstraint<'db>) -> Option<Self> {
-        self.hole
-            .intersect(db, positive)
-            .map(|hole| NegatedRangeConstraint { hole })
-    }
-
-    /// Returns the union of two negative constraints. (This this is _intersection_ of the
-    /// constraints' holes.)
-    fn union_negative(
-        &self,
-        db: &'db dyn Db,
-        positive: &NegatedRangeConstraint<'db>,
-    ) -> Option<Self> {
-        self.hole
-            .intersect(db, &positive.hole)
-            .map(|hole| NegatedRangeConstraint { hole })
-    }
-
-    /// Returns the intersection of two negative constraints. (This this is _union_ of the
-    /// constraints' holes.)
-    fn intersect_negative(
-        &self,
-        db: &'db dyn Db,
-        other: &NegatedRangeConstraint<'db>,
-    ) -> Option<Self> {
-        self.hole
-            .union(db, &other.hole)
-            .map(|hole| NegatedRangeConstraint { hole })
+impl<'db> SatisfiedClauses<'db> {
+    fn push(&mut self, clause: SatisfiedClause<'db>) {
+        self.clauses.push(clause);
     }
 
-    fn display(&self, db: &'db dyn Db, typevar: impl Display) -> impl Display {
-        struct DisplayNegatedRangeConstraint<'a, 'db, D> {
-            constraint: &'a NegatedRangeConstraint<'db>,
-            typevar: D,
-            db: &'db dyn Db,
-        }
-
-        impl<D: Display> Display for DisplayNegatedRangeConstraint<'_, '_, D> {
-            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-                if (self.constraint.hole.lower)
-                    .is_equivalent_to(self.db, self.constraint.hole.upper)
-                {
-                    return write!(
-                        f,
-                        "({} ≠ {})",
-                        &self.typevar,
-                        self.constraint.hole.lower.display(self.db)
-                    );
+    /// Simplifies the DNF representation, removing redundancies that do not change the underlying
+    /// function. (This is used when displaying a BDD, to make sure that the representation that we
+    /// show is as simple as possible while still producing the same results.)
+    fn simplify(&mut self) {
+        while self.simplify_one_round() {
+            // Keep going
+        }
+    }
```

Comment:
Is it not possible to sufficiently simplify the underlying BDD so we don't need additional simplification of the `SatisfiedConstraints`?

Reply:
I see, makes sense. It still somehow feels like we ought to be able to do this "on the fly" while building the DNF representation, by tracking the information we can assume and not adding clauses that can be assumed? But maybe not, or maybe doing it after-the-fact is just as good in practice.